File: section-03-basic-maths/10-check-for-prime-number/check-for-prime-number-opt.cs
using System;

class Solution
{
    /* Function to find whether the
    number is prime or not */
    public bool IsPrime(int n)
    {
        // Edge case
        if (n < 2) return false;

        // Loop from 2 to sqrt(n)
        for (int i = 2; i * i <= n; ++i)
        {
            // Check if i is a divisor
            if (n % i == 0)
            {
                return false;
            }
        }

        // Return true as the number is prime
        return true;
    }
}




/* 

Time Complexity: O(sqrt(N)) – Looping sqrt(N) times to find the count of all divisors of N.

Space Complexity: O(1) – Using a couple of variables i.e., constant space.

 */

File: section-03-basic-maths/10-check-for-prime-number/check-for-prime-number-opt.py
class Solution:
    # Function to find whether the
    # number is prime or not
    def isPrime(self, n):
        # Edge case
        if n < 2:
            return False

        # Loop from 2 to √n
        i = 2
        while i * i <= n:
            # Check if i is a divisor
            if n % i == 0:
                return False
            i += 1

        # Return true as the number is prime
        return True

if __name__ == "__main__":
    n = 5

    # Creating an instance of 
    # Solution class
    sol = Solution()

    # Function call to find whether the
    # given number is prime or not
    ans = sol.isPrime(n)

    if ans:
        print(f"{n} is a prime number.")
    else:
        print(f"{n} is not a prime number.")







""" 
For Loop approach 

import math


class Solution:
    def isPrime(self, n):
        if n < 2:
            return False

        for i in range(2, int(math.sqrt(n)) + 1):
            if n % i == 0:
                return False

        return True

        
"""